pkg/volume/csi/csi_block.go

```diff
@@ -59,17 +58,15 @@ func (m *csiBlockMapper) GetGlobalMapPath(spec *volume.Spec) (string, error) {
 }
```

Comment:
globalMapPath should be based on volumeID too, not pv name.

Reply:
@wongma7 

I tried to fix both `getVolumeDevicePluginDir` and `getVolumeDeviceDataDir` to be based on `volumeID`. However, in `NewBlockVolumeUnmapper`, `volumeID` is gotten by calling `loadVolumeData` with specifying `unmapper.specName`, or volName. So, at least, `getVolumeDeviceDataDir` won't be able to be based on `volumeID`.

Should we make only `getVolumeDevicePluginDir` based on `volumeID`, or keep the paths as they are?
I think that pv name and volumeID is 1:1, so this change won't be a mandatory. Then, we would be easier to debug the volume if we keep device and volume data under the same directory.

WDYT?

Reply:
Why is pv name and volumeID 1:1? Can't PV name be anything? I assume that whoever wrote this code had a good reason to think pvname == volumeID but I am not so sure. I thought volumeID is supposed to be 1:1 with volumeHandle, not PV name?

Reply:
It seems like variables are poorly named. "specVolID" is PV name. https://github.com/kubernetes/kubernetes/blob/39724859b524e345881eb8ef458f0d5a2f25bd9a/pkg/volume/csi/csi_mounter.go#L45 . This is not the same thing as `volume_id` in the CSI spec, that is volumeHandle?!

Reply:
The issue I see is it's possible (but unlikely) for the same volume_id to be used by two different PVs, in which case there will be two globalMapPaths for the same `volume_id` meaning NodeStage may be called early.

I think you can keep the paths as they are to make this PR simpler, and you can leave the sorting out of paths as a follow-up for me...

Reply:
@wongma7 

It seems that we have been using [`spec.Name()`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L472) as an ID for volume in [GetPodDeviceMapPath](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/gce_pd_block.go#L178) in in-tree plugins. So, we should be able to use `m.specName` as CSI plugin's ID for volume.

The value of `m.specName` is like `pvc-82051086-eb98-4dd0-941a-1512bc438278`, so we won't need to sanitize it by calling `utilstrings.EscapeQualifiedName`.
Also, we should change the comment from `{pvname}` to `{specName}` to avoid confusion.

If you agree. I will change as so.

Reply:
I agree 👍

Reply:
While I am not enough of a linux pro to understand why it works, I tried the fd lock thing today and it works as described by the comment before AttachFileDevice. I will need to check if it still works for the new bind mounted case...

1. Start some nonsense container that periodically writes to the device
docker run --name test --device /dev/nvme2n3 nginx bash -c 'i=0; while true; do echo $i; ((i++)); echo $i > ./hello.txt; dd if=./hello.txt of=/dev/nvme2n3 bs=64 count=1; sleep 5; done'; docker rm test
2. Detach the device, then it will get:
dd: failed to open '/dev/nvme2n3': No such device or address
3. Attach a different device, it will resume writing to it. This is the problem the comment for the lock described.

If you setup a loop device to keep the fd open:

1. docker run --name test --device /dev/nvme2n3 nginx bash -c 'i=0; while true; do echo $i; ((i++)); echo $i > ./hello.txt; dd if=./hello.txt of=/dev/nvme2n3 bs=64 count=1; sleep 5; done'; docker rm test
2. sudo losetup -f --show /dev/nvme2n3
3. Detach the device, then it will get:
dd: failed to open '/dev/nvme2n3': No such device or address
4. Attach a different device, it will continue to get the above error.